# 🔍 Credit Card Fraud Detection

## Project Outline

| Phase | Description |
|-------|-------------|
| 1. Problem Definition | What we're solving and why |
| 2. Data Collection | Dataset from Kaggle (284,807 transactions) |
| 3. Data Cleaning | Handling missing values, fixing data types |
| 4. EDA | Understanding data, visualizing fraud vs legit |
| 5. Feature Engineering | Scaling, new features, handling imbalance (SMOTE) |
| 6. Model Building | Using Machine Learning — Logistic Regression |
| 7. Model Testing | Testing model on unseen data, train/test split |
| 8. Evaluation | Precision, Recall, F1, ROC-AUC |
| 9. Conclusion | Best model, ethical considerations |

**Dataset:** [Credit Card Fraud Detection — Kaggle](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud)  
**Target:** Predict fraudulent transactions (Class: 0 = Legit, 1 = Fraud)

## Phase 1 — Problem Definition

Credit Card Fraud Detection is a binary classification problem where we predict 
whether a transaction is fraudulent (1) or legitimate (0).

Credit card fraud is a major financial threat causing billions of dollars in losses 
annually. Detecting fraud manually is impossible due to the sheer volume of 
transactions happening every second. This is where Machine Learning helps and can 
automatically detect suspicious patterns in transaction data.

The dataset is highly imbalanced -only 0.17% of transactions are fraud. This means 
a model that predicts everything as legitimate would still be 99.83% accurate, which 
is misleading. This is why we focus on Precision, Recall and F1 Score rather than 
plain accuracy.

Our goal is to build a Logistic Regression model that can correctly identify 
fraudulent transactions while minimizing false alarms.

In [5]:
# numpy - numerical and mathematical computations
import numpy as np

#data loading and manipulation
import pandas as pd

# data visualization
import matplotlib.pyplot as plt
import seaborn as sns

#  machine learning tools
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# SMOTE - handling imbalanced data
from imblearn.over_sampling import SMOTE

## Phase 2 — Data Collection

The dataset is sourced from Kaggle — Credit Card Fraud Detection.
It contains 284,807 transactions made by European cardholders in September 2013.
Published by: Machine Learning Group, (ULB).

In [ ]:
# loading the dataset
df = pd.read_csv('data/creditcard.csv')

# data's characterstics
print("Shape:", df.shape)
print("Columns:", df.columns.tolist())
df.head()

Shape: (284807, 31)
Columns: ['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount', 'Class']


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
284802,172786.0,-11.881118,10.071785,-9.834783,-2.066656,-5.364473,-2.606837,-4.918215,7.305334,1.914428,...,0.213454,0.111864,1.014480,-0.509348,1.436807,0.250034,0.943651,0.823731,0.77,0
284803,172787.0,-0.732789,-0.055080,2.035030,-0.738589,0.868229,1.058415,0.024330,0.294869,0.584800,...,0.214205,0.924384,0.012463,-1.016226,-0.606624,-0.395255,0.068472,-0.053527,24.79,0
284804,172788.0,1.919565,-0.301254,-3.249640,-0.557828,2.630515,3.031260,-0.296827,0.708417,0.432454,...,0.232045,0.578229,-0.037501,0.640134,0.265745,-0.087371,0.004455,-0.026561,67.88,0
284805,172788.0,-0.240440,0.530483,0.702510,0.689799,-0.377961,0.623708,-0.686180,0.679145,0.392087,...,0.265245,0.800049,-0.163298,0.123205,-0.569159,0.546668,0.108821,0.104533,10.00,0
284806,172792.0,-0.533413,-0.189733,0.703337,-0.506271,-0.012546,-0.649617,1.577006,-0.414650,0.486180,...,0.261057,0.643078,0.376777,0.008797,-0.473649,-0.818267,-0.002415,0.013649,217.00,0


## Phase 3 — Data Cleaning
Checking for missing values, duplicate rows and correct data types.

In [12]:
print("Missing Values:\n", df.isnull().sum())
# for missing values use mean for evenly distributed numerical data ,median for noticable outliers and mode for strings
# if missing valued rows is < 5% ,remove the row using df.dropna()
print("\nDuplicate Rows:", df.duplicated().sum())
# to remove duplicates using df.drop_duplicates()
print("\nData Types:\n",  all(df.dtypes != 'object'))
# use type casting for primitive data types , methods such as lower, strip , replace for typos ,toDate for dates

Missing Values:
 Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Duplicate Rows: 0

Data Types:
 True


In [13]:
df = df.drop_duplicates()
print("Duplicates removed! New shape:", df.shape)

Duplicates removed! New shape: (283726, 31)


### Data Cleaning Results

- Missing Values: 0 — No missing values found, no action needed.
- Duplicate Rows: 1081 — Duplicates found and removed. New shape: (283726, 31)
- Data Types: All columns are numerical (float64/int64), no text columns found.

If any column had string/text data type, we would have used One Hot Encoding to 
convert them into numerical columns since Machine Learning models can only work 
with numbers.

Dataset is now clean and ready for Exploratory Data Analysis.